libs/core/src/lib/select/commons/mat-select-commons.ts

```diff
@@ -0,0 +1,917 @@
+
+            // Scrolls the panel up by the distance it was extending past the boundary, then
+            // adjusts the offset by that amount to move the panel up into the viewport.
+            this._scrollTop -= distanceBelowViewport;
+            this._offsetY -= distanceBelowViewport;
+            this._transformOrigin = this._getOriginBasedOnOption();
+
+            // If the panel is scrolled to the very top, it won't be able to fit the panel
+            // by scrolling, so set the offset to 0 to allow the fallback position to take
+            // effect.
+            if (this._scrollTop <= 0) {
+                this._scrollTop = 0;
+                this._offsetY = 0;
+                this._transformOrigin = `50% bottom 0px`;
+            }
+        }
+
+        /**
+         * @hidden
+         */
+        private _adjustPanelDown(panelHeightTop: number, topSpaceAvailable: number,
+                                 maxScroll: number) {
+            // Browsers ignore fractional scroll offsets, so we need to round.
+            const distanceAboveViewport = Math.round(panelHeightTop - topSpaceAvailable);
+
+            // Scrolls the panel down by the distance it was extending past the boundary, then
+            // adjusts the offset by that amount to move the panel down into the viewport.
+            this._scrollTop += distanceAboveViewport;
+            this._offsetY += distanceAboveViewport;
+            this._transformOrigin = this._getOriginBasedOnOption();
+
+            // If the panel is scrolled to the very bottom, it won't be able to fit the
+            // panel by scrolling, so set the offset to 0 to allow the fallback position
+            // to take effect.
+            if (this._scrollTop >= maxScroll) {
+                this._scrollTop = maxScroll;
+                this._offsetY = 0;
+                this._transformOrigin = `50% top 0px`;
+                return;
+            }
+        }
+
+
+        /**
+         * Sets the transform origin point based on the selected option.
+         *
+         * @hidden
+         */
+        private _getOriginBasedOnOption(): string {
+            const itemHeight = this._getItemHeight();
+            const optionHeightAdjustment = (itemHeight - this._triggerRect.height) / 2;
+            const originY = Math.abs(this._offsetY) - optionHeightAdjustment + itemHeight / 2;
+            return `50% ${originY}px 0px`;
+        }
+
+
+        /**
+         *
+         * @hidden
+         */
+        private _highlightCorrectOption(): void {
+            if (this._keyManager) {
+                if (this.empty) {
+                    this._keyManager.setFirstItemActive();
+                } else {
+                    this._keyManager.setActiveItem(this.selected);
+                }
+            }
+        }
+
+        /** Handles keyboard events while the select is closed. */
+        private _handleClosedKeydown(event: KeyboardEvent): void {
+            const keyCode = event.keyCode;
```

Comment:
Should not rely on `keyCode`

Reply:
We are slowly replacing our `keyCode` usages with `KeyUtil.isKey` utility, but I think we already talked about this during last call.

Reply:
let me double check.. i have replaced all keycodes and updated KeyUtils